// sketch-constants/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var FillType = exports.FillType = {
    Solid: 0,
    Gradient: 1,
    Pattern: 4,
    Noise: 5
};

var GradientType = exports.GradientType = {
    Linear: 0,
    Radial: 1,
    Circular: 2
};

var PatternFillType = exports.PatternFillType = {
    Tile: 0,
    Fill: 1
};

var NoiseFillType = exports.NoiseFillType = {
    Original: 0,
    Black: 1,
    White: 2,
    Color: 3
};

// _MSStyleBorderOptions
var BorderLineCapsStyle = exports.BorderLineCapsStyle = {
    Butt: 0,
    Round: 1,
    Square: 2
};

var BorderLineJoinStyle = exports.BorderLineJoinStyle = {
    Miter: 0,
    Round: 1,
    Bevel: 2
};

var LineDecorationType = exports.LineDecorationType = {
    None: 0,
    OpenedArrow: 1,
    ClosedArrow: 2,
    Bar: 3
};

// _MSStyleBlur
var BlurType = exports.BlurType = {
    GaussianBlur: 0,
    MotionBlur: 1,
    ZoomBlur: 2,
    BackgroundBlur: 3
};

var BorderPosition = exports.BorderPosition = {
    Center: 0,
    Inside: 1,
    Outside: 2,
    Both: 3 // This is Sketch internal option - don't use it.
};

var MaskMode = exports.MaskMode = {
    Outline: 0,
    Alpha: 1
};

var BooleanOperation = exports.BooleanOperation = {
    None: -1,
    Union: 0,
    Subtract: 1,
    Intersect: 2,
    Difference: 3
};

var ExportOptionsFormat = exports.ExportOptionsFormat = {
    PNG: 'png',
    JPG: 'jpg',
    TIFF: 'tiff',
    PDF: 'pdf',
    EPS: 'eps',
    SVG: 'svg'
};

var BlendingMode = exports.BlendingMode = {
    Normal: 0,
    Darken: 1,
    Multiply: 2,
    ColorBurn: 3,
    Lighten: 4,
    Screen: 5,
    ColorDodge: 6,
    Overlay: 7,
    SoftLight: 8,
    HardLight: 9,
    Difference: 10,
    Exclusion: 11,
    Hue: 12,
    Saturation: 13,
    Color: 14,
    Luminosity: 15
};

var TextAlignment = exports.TextAlignment = {
    Left: 0,
    Right: 1,
    Center: 2,
    Justified: 3
};

var TextBehaviour = exports.TextBehaviour = {
    Auto: 0,
    Fixed: 1
};

var CurvePointMode = exports.CurvePointMode = {
    Straight: 1,
    Mirrored: 2,
    Disconnected: 4,
    Asymmetric: 3
};

var ModelClasses = exports.ModelClasses = {
    // Base model classes.
    MSModelObject: 'MSModelObject',
    MSModelBase: 'MSModelBase',
    MSDocumentData: 'MSDocumentData',

    // Base layers.
    MSLayer: 'MSLayer',

    // Core layers.
    MSShapeGroup: 'MSShapeGroup',
    MSLayerGroup: 'MSLayerGroup',
    MSPage: 'MSPage',
    MSTextLayer: 'MSTextLayer',
    MSBitmapLayer: 'MSBitmapLayer',
    MSSliceLayer: 'MSSliceLayer',
    MSArtboardGroup: 'MSArtboardGroup',

    // Primitive shapes.
    MSOvalShape: 'MSOvalShape',
    MSRectangleShape: 'MSRectangleShape',
    MSStarShape: 'MSStarShape',
    MSPolygonShape: 'MSPolygonShape',
    MSTriangleShape: 'MSTriangleShape',

    // Style related.
    MSStyle: 'MSStyle',
    MSStyleBorder: 'MSStyleBorder',
    MSStyleFill: 'MSStyleFill',
    MSStyleShadow: 'MSStyleShadow',
    MSStyleInnerShadow: 'MSStyleInnerShadow',
    MSStyleBlur: 'MSStyleBlur',
    MSStyleBorderOptions: 'MSStyleBorderOptions',
    MSStyleColorControls: 'MSStyleColorControls',
    MSStyleReflection: 'MSStyleReflection'
};
